python_modules/dagster/dagster_tests/general_tests/grpc_tests/test_metrics.py

```diff
@@ -0,0 +1,108 @@
+import inspect
+import json
+import re
+import subprocess
+import threading
+import time
+from typing import Any
+
+from dagster._core.host_representation import ExternalRepositoryOrigin
+from dagster._core.host_representation.origin import (
+    GrpcServerCodeLocationOrigin,
+)
+from dagster._core.instance import DagsterInstance
+from dagster._core.test_utils import (
+    instance_for_test,
+)
+from dagster._grpc.client import DagsterGrpcClient
+from dagster._grpc.server import (
+    METRICS_RETRIEVAL_FUNCTIONS,
+    DagsterApiServer,
+    wait_for_grpc_server,
+)
+from dagster._grpc.types import SensorExecutionArgs
+from dagster._serdes import deserialize_value
+from dagster._utils import (
+    file_relative_path,
+    find_free_port,
+)
+
+
+def is_grpc_method(attr_name: str, obj: Any) -> bool:
+    """Grpc methods use pascal casing. This function checks if the given attribute is a method with pascal casing."""
+    pascal_case_rgx = re.compile(r"^[A-Z][a-z]*([A-Z][a-z]*)")
+    return (
+        inspect.isfunction(getattr(obj, attr_name)) and pascal_case_rgx.match(attr_name) is not None
+    )
+
+
+def test_metrics_retrieval_annotations():
+    for attr in dir(DagsterApiServer):
+        if is_grpc_method(attr, DagsterApiServer):
+            if attr != "ExternalSensorExecution":
+                continue
+            assert attr in METRICS_RETRIEVAL_FUNCTIONS, "Missing annotation for method: " + attr
+    for attr in METRICS_RETRIEVAL_FUNCTIONS:
+        if attr != "ExternalSensorExecution":
+            continue
+        assert is_grpc_method(attr, DagsterApiServer), "Missing method for annotation: " + attr
+
+
+def _launch_sensor_execution(
+    repo_origin: ExternalRepositoryOrigin, client: DagsterGrpcClient, instance: DagsterInstance
+):
+    client.external_sensor_execution(
+        sensor_execution_args=SensorExecutionArgs(
+            repository_origin=repo_origin,
+            instance_ref=instance.get_ref(),
+            sensor_name="extremely_slow_sensor",
+            last_tick_completion_time=None,
+            last_run_key=None,
+            cursor=None,
+            timeout=2,
+            last_sensor_start_time=None,
+        ),
+    )
+
+
+def test_ping_metrics_retrieval():
+    port = find_free_port()
```

Comment:
this is a more detailed example of what it's actually like to retrieve metrics.
1. In another thread, we launch a sensor execution on the grpc server.
2. We wait to ensure that the sensor execution is occurring
3. We retrieve metrics, and expect to see that ongoing sensor execution represented.